Quick Pickup QuickMenu

RayureBob opened the pull request:
Addition of a feature listing potentially interesting items within current zone, in a QuickMenu with options.
Upon confirming a selection of one or more item, the player character goes picking them up.

It also adds a settings screen allowing customization or items displayed in the popup.

RayureBob committed:
Added Quick Pickup Menu and its settings screen

RayureBob committed:
Perfected Quick Pickup settings screen and quickmenu display + settings serialization

egocarib commented on Harmony Patches/Patch_XRL_World_GameObject_FireEvent.cs:
```diff
@@ -0,0 +1,22 @@
+public class Patch_XRL_World_GameObject_FireEvent
+{
+    [HarmonyPatch("FireEvent", typeof(Event)), HarmonyPrefix]
+    static void ShouldAutoget(GameObject __instance, Event E)
```
Minor thing, but function should probably be named differently here (ex: "Prefix" or "FireEvent" or something more relevant)

egocarib commented on Harmony Patches/Patch_XRL_World_GameObject_FireEvent.cs:
```diff
@@ -0,0 +1,22 @@
+            GameObject g = E.GetParameter("Object") as GameObject;
+            if(g == null) return;
+
+            if(g.GetStringProperty(Constants.QuickPickupProperty) == "yes")
```
If the string itself doesn't have any significance, might make more sense to use GetIntProperty/SetIntProperty/RemoveIntProperty.

egocarib commented on Harmony Patches/Patch_XRL_World_GameObject_FireEvent.cs:
```diff
@@ -0,0 +1,22 @@
+    [HarmonyPatch("FireEvent", typeof(Event)), HarmonyPrefix]
+    static void ShouldAutoget(GameObject __instance, Event E)
+    {
+        if(E.ID == "CommandTakeObject")
```
Is there a reason this requires a Harmony patch, instead of simply registering for the `CommandTakeObject` event within `QudUX_QuickPickupPart.Register()`? I guess this code would run when _any_ creature picks up the item, versus registering the event would have it run only when the player picks up the item. Maybe that's a significant difference here?

egocarib commented on Screens/QudUX_QuickPickupSettingsScreen.cs:
```diff
@@ -0,0 +1,357 @@
+            { "Rifles",             "&yAll {{W|Rifles}} of the game"},
+            { "Heavy Weapons",      "&yAll {{W|Heavy Weapons}} of the game"},
+            { "Bows",               "&yAll {{W|Bows}} of the game"},
+            { "Shields",            "&yAll {{W|Shields}} of the game"},
```
I would recommend removing "of the game" from the end of all of these descriptions. I don't think it's adding much. the descriptions feel simpler to understand without it IMO

egocarib commented on Screens/QudUX_QuickPickupSettingsScreen.cs:
```diff
@@ -0,0 +1,357 @@
+        {
+            { "Tier 0",             "&yEvery {{w|Bronze}} item"},
+            { "Tier 1",             "&y{{W|Short Bow}} and least powerful {{W|Iron}} equipment and weapons"},
+            { "Tier 2",             "&y{{W|Revolver}}, {{W|Musket}}, {{W|Compound Bow}}... And steel {{W|Steel}} equipment and weapons"},
```
the word "steel" is repeated twice on this line.

egocarib commented:
This is a very cool feature, and I like how you've set up the screen! nice work.

I left a few code review comments above.

One other general thought:
I noticed that my character also does _other_ auto-explore things mixed in with quick pickup. For example, I selected only one item to quick pick up, but after I confirmed in the menu, my character walked around the screen and also looked at two other painted items before picking up the item I had selected. I think this is probably OK if it's not easy to separate from the auto-explore behavior, but I thought I'd call it out. One simple thing that might make this easier to understand could be to add a note to one of the menus that says "quick pickup may also perform other auto-explore actions" or something like that. But I'll let you decide if you think that's worth adding or not. If not, I'm also okay with how it works now.

RayureBob committed:
Code review changes

RayureBob committed:
Upgraded item sorting + modified feature description in screen

RayureBob committed:
Fixed pronouns in sparafucile's restocker dialogs + modified QuickPickup feature description on its settings screen

RayureBob committed:
cleaning

RayureBob commented:
I've fixed everything, and changed the description of the feature on its dedicated screen, but I'm not satisfied by the fact it relies on auto-explore. It hasn't been made for that. There's a class one can extend to use AutoAct, OngoingAction, which is the base class of dissassembly for instance. I'm going to try to rework the feature so that it delegates the behaviour to AutoAct through an OngoinAction object. It would be a much cleaner and logical solution

RayureBob committed:
Integrated feature through AutoAct Extension system

RayureBob committed:
Modified screen to fit rework

RayureBob commented:
It's done, much cleaner version, without having the character wander here and there along its way.
